src/make_watertight/CheckWatertight.cpp

```diff
@@ -0,0 +1,391 @@
+  void* input_dim[] = {&dim};
+  result = MBI()->get_entities_by_type_and_tag(input_set, moab::MBENTITYSET, &geom_tag,
+                                               input_dim, 1, surf_sets);
+  if (moab::MB_SUCCESS != result) {
+    return result;
+  }
+
+  // volume sets
+  dim = 3;
+  result = MBI()->get_entities_by_type_and_tag(input_set, moab::MBENTITYSET, &geom_tag,
+                                               input_dim, 1, vol_sets);
+  if (moab::MB_SUCCESS != result) {
+    return result;
+  }
+  if (!test) {
+    std::cout << "number of surfaces=" << surf_sets.size() << std::endl;
+    std::cout << "number of volumes="  << vol_sets.size() << std::endl;
+  }
+  // ******************************************************************
+  // Watertightness is a property of volumes. Check each surface for
+  // watertightness.
+  // ******************************************************************
+  // counted leaky surfaces
+  int total_counter = 0, unmatched_counter = 0;
+  std::set<int> leaky_surfs, leaky_vols;
+  moab::Skinner tool(MBI());
+
+  // remove all edges for fast skinning
+  moab::Range edges;
+
+  result = MBI()->get_entities_by_type(0, moab::MBEDGE, edges);   // get all edges
+  if (moab::MB_SUCCESS != result) { // failed to get edge data
+    return result; // failed
+  }
+//10/11/13
+//removed as a result of the change from the gen::find_skin function to the moab::Skinner:find_skin function
+  /*
+    result = MBI()->delete_entities( edges ); //otherwise delete all edge
+
+    if(moab::MB_SUCCESS != result) // failed to delete edge data
+      {
+        return result; // failed
+      }
+  */
+  // loop over each volume meshset
+  int vol_counter = 0;
+  for (moab::Range::iterator i = vol_sets.begin(); i != vol_sets.end(); ++i) {
+    ++vol_counter;
+    int surf_counter = 0;
+    moab::Range child_sets;
+
+    result = MBI()->get_child_meshsets(*i, child_sets);   // get child set
+    if (moab::MB_SUCCESS != result) {
+      return result; // failed
+    }
+
+    // get the volume id of the volume meshset to print a status message
+    int vol_id = 0;
+    // i is the iterator, so &(*i) is a pointer to the first element of moab::Range
+    result = MBI()->tag_get_data(id_tag, &(*i), 1, &vol_id);
+
+    if (moab::MB_SUCCESS != result) {
+      return result;
+    }
+
+    if (verbose) {
+      std::cout << "checking volume " << vol_counter << "/" << vol_sets.size()
+                << " id=" << vol_id << std::endl;
+    }
+
+    // determine how many skin edges are in each volume
+    int n_tris = 0;
+
+    for (moab::Range::iterator j = child_sets.begin(); j != child_sets.end(); ++j) {
+      result = MBI()->get_number_entities_by_type(*j, moab::MBTRI, n_tris);   // for each child set get number of triangles
+      if (moab::MB_SUCCESS != result) {
+        return result;
+      }
+    }
+
+    // save the edges in a vector that is large enough to avoid resizing
+    // presumably some kind of efficiency thing?? ad ??
```

Comment:
Yep, it is an efficiency thing.  C++ STD vectors are stored in arrays.  When a vector needs more capacity, a new array is allocated (typically 1.5x to 2x the size [1]), and the contents of the old array are copied to the new array.

If you know upfront that you need a large array, this can save significant compute overhead.

[1] https://stackoverflow.com/a/7397862/1015951